src/commonMain/kotlin/fr/acinq/lightning/channel/InteractiveTx.kt

```diff
@@ -341,23 +396,80 @@ data class SharedTransaction(
         return Transaction(2, inputs, outputs, lockTime)
     }
 
-    fun sign(keyManager: KeyManager, fundingParams: InteractiveTxParams, localParams: LocalParams, remoteNodeId: PublicKey): PartiallySignedSharedTransaction {
+    fun sign(session: InteractiveTxSession, keyManager: KeyManager, fundingParams: InteractiveTxParams, localParams: LocalParams, remoteNodeId: PublicKey): PartiallySignedSharedTransaction {
         val unsignedTx = buildUnsignedTx()
         val sharedSig = fundingParams.sharedInput?.sign(keyManager.channelKeys(localParams.fundingKeyPath), unsignedTx)
+        val sharedOutput = fundingParams.sharedInput?.let { i -> mapOf(i.info.outPoint to i.info.txOut) } ?: mapOf()
+        val localOutputs = localInputs.associate { i -> i.outPoint to i.txOut }
+        val remoteOutputs = remoteInputs.associate { i -> i.outPoint to i.txOut }
+        val previousOutputsMap = sharedOutput + localOutputs + remoteOutputs
+        val previousOutputs = unsignedTx.txIn.map { previousOutputsMap[it.outPoint]!! }.toList()
+
+        // nonces that we've received for all musig2 swap-in
+        val receivedNonces: Map<Long, IndividualNonce> = when (session.txCompleteReceived) {
```

Comment:
This can be simplified to use the same pattern as eclair:

- validate (once) that `tx_complete` contains the right number of nonces
- zip them with our sorted musig2 remote inputs (without redundant validation)

See the corresponding comment on the eclair PR.

Reply:
I see what you mean, it becomes messy to untangle the local and remote swap inputs, this map is the simplest way of writing that code, let's forget about this.